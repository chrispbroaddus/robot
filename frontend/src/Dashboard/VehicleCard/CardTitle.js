import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

const Container = styled.div`margin: 10px;`

const Title = styled.span`
  color: #90a4ae;
  font-family: Montserrat-Light;
  font-size: 10px;
  letter-spacing: 0.5px;
  text-align: left;
`

const TitleCard = props => {
  return (
    <Container>
      <Title>
        {props.title}
      </Title>
    </Container>
  )
}

TitleCard.propTypes = {
  title: PropTypes.string.isRequired
}

export default TitleCard
