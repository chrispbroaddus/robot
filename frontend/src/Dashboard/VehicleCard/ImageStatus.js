import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { VEHICLE_STATUS } from '../../redux/modules/vehicles'

const Container = styled.div`
  position: relative;
  padding: 0;
`

const Status = styled.div`
  position: absolute;
  z-index: 1;
  bottom: 10px;
  right: 12px;
  padding: 5px;
  font-size: 9px;
  line-height: 11px;
  text-align: center;
  text-transform: uppercase;
  border-radius: 2px;
  background-color: rgba(54, 54, 54, 0.5);
  color: ${props => props.color};
`

const Image = styled.img`
  width: 265px;
  position: relative;
  z-index: 0;
`

const ImageStatus = props => {
  let color
  let status
  if (props.status === VEHICLE_STATUS.INACTIVE) {
    color = '#fe7867'
    status = 'Inactive'
  } else if (props.status === VEHICLE_STATUS.ACTIVE) {
    color = '#50e3c2'
    status = VEHICLE_STATUS.ACTIVE
  } else if (
    props.status === VEHICLE_STATUS.OFFLINE ||
    props.status === 'launching'
  ) {
    color = '#f65045'
    status = 'Offline'
  } else if (props.status === VEHICLE_STATUS.NETWORK_DROP) {
    color = '#f65045'
    status = 'Unavailable'
  } else if (props.status === VEHICLE_STATUS.DISABLED) {
    color = 'lightslategray'
    status = 'Disabled'
  }

  return (
    <Container>
      <Status color={color}>
        {status}
      </Status>
      <Image src={props.src} alt="last vehicle frame" />
    </Container>
  )
}

ImageStatus.propTypes = {
  status: PropTypes.string.isRequired,
  src: PropTypes.string.isRequired,
  title: PropTypes.string
}

export default ImageStatus
