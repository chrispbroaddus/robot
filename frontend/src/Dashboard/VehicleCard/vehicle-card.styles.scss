@import '~@Themes/_variables.scss';

.card {
  width: 265px;
  height: fit-content;
  border-radius: 2px;
  background-color: #fff;
  box-shadow: -1px 2px 8px 0 rgba(155, 155, 155, 0.2);
  padding-bottom: 17px;
  margin-right: 40px;
  margin-bottom: 40px;
  display: flex;
  flex-flow: column nowrap;
  transition: 0.2s all;
}

.wait-text {
  color: lightslategray;
  text-align: left;
  letter-spacing: 0;
  font-size: 10px;
  padding: 10px;
  margin-top: 0;
}

.wait-text::before {
  animation: pulse 0.5s ease-in infinite;
  font-family: 'Montserrat-Medium';
  content: '|';
  margin-right: 10px;
  color: $red;
}

@keyframes pulse {
  0% {
    opacity: 1;
  }

  50% {
    opacity: 0.2;
  }

  100% {
    opacity: 1;
  }
}
