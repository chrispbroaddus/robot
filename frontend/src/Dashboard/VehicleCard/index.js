import React from 'react'
import PropTypes from 'prop-types'
import './vehicle-card.styles.scss'
import DefaultImage from '../../images/default.png'
import CardTitle from './CardTitle'
import ImageStatus from './ImageStatus'
import DashboardButton from '../DashboardButton'
import { VEHICLE_STATUS, VEHICLE_TYPES } from '../../redux/modules/vehicles'

class VehicleCard extends React.Component {
  static propTypes = {
    vehicleId: PropTypes.string.isRequired,
    handleClickOperate: PropTypes.func.isRequired,
    handleClickShutdown: PropTypes.func.isRequired,
    cameraRole: PropTypes.string,
    status: PropTypes.string,
    type: PropTypes.string,
    realVehicleId: PropTypes.string.isRequired,
    getLatestCameraFrame: PropTypes.func.isRequired
  }

  static defaultProps = {
    cameraRole: '',
    status: '',
    type: ''
  }

  constructor(props) {
    super(props)

    this.state = {
      imageUrl: null
    }
  }

  componentWillMount() {
    const { getLatestCameraFrame, realVehicleId } = this.props
    getLatestCameraFrame(realVehicleId, 'FrontFisheye')
      .then(data => {
        this.setState({
          imageUrl: data.url
        })
      })
      .catch(() => {
        this.setState({
          imageUrl: DefaultImage
        })
      })
  }

  render() {
    const {
      vehicleId,
      cameraRole,
      status,
      type,
      handleClickOperate,
      handleClickShutdown
    } = this.props
    let { imageUrl } = this.state
    imageUrl = imageUrl ? imageUrl : DefaultImage

    return (
      <div className="card">
        <ImageStatus src={imageUrl} status={status} title={vehicleId} />
        <CardTitle title={vehicleId} />
        <div className="center">
          {status === VEHICLE_STATUS.NETWORK_DROP && (
            <p className="wait-text">
              Connection difficulties. Try refreshing.
            </p>
          )}
          {status === VEHICLE_STATUS.LAUNCHING && (
            <p className="wait-text">Launching Simulator . . .</p>
          )}
          {status === VEHICLE_STATUS.ACTIVE && (
            <DashboardButton
              buttonClassName={status}
              disabled={status === VEHICLE_STATUS.DISABLED}
              onClick={() => handleClickOperate(vehicleId, cameraRole)}
            >
              operate
            </DashboardButton>
          )}
          {type === VEHICLE_TYPES.SIMULATOR && (
            <DashboardButton
              buttonClassName={status}
              disabled={status === VEHICLE_STATUS.DISABLED}
              onClick={() => handleClickShutdown(vehicleId)}
            >
              shut down
            </DashboardButton>
          )}
        </div>
      </div>
    )
  }
}

export default VehicleCard
