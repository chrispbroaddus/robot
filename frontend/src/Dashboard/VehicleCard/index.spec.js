import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import VehicleCard from './index'

describe('VehicleCard', () => {
  let props

  beforeEach(() => {
    props = {
      cameraRole: '',
      vehicleId: '',
      realVehicleId: '',
      status: '',
      type: '',
      handleClickOperate: () => {},
      handleClickShutdown: () => {},
      getLatestCameraFrame: () => {
        return new Promise(() => {})
      }
    }
  })

  it('should render self and subcomponents', () => {
    const wrapper = shallow(<VehicleCard {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
