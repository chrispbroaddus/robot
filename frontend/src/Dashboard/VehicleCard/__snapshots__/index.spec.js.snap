// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleCard should render self and subcomponents 1`] = `
<div
  className="card"
>
  <ImageStatus
    src="default.png"
    status=""
    title=""
  />
  <TitleCard
    title=""
  />
  <div
    className="center"
  />
</div>
`;
